server/src/main/java/io/mantisrx/server/master/SchedulingService.java

```diff
@@ -330,28 +333,66 @@ private TaskSchedulingService setupTaskSchedulingService(TaskScheduler taskSched
         return hasValue ? Optional.of(lease.getAttributeMap().get(attributeName).getText().getValue()) : Optional.empty();
     }
 
+    /**
+     * Attempts to launch tasks given some number of leases from Mesos.
+     *
+     * When a task is launched successfully, the following will happen:
+     *
+     * 1. Emit a {@link WorkerLaunched} event to be handled by the corresponding actor.
+     * 2. Makes a call to the underlying Mesos driver to launch the task.
+     *
+     * A task can fail to launch if:
+     *
+     * 1. It doesn't receive enough metadata for {@link WorkerPorts} to pass its preconditions.
+     *      - No launch task request will be made for this assignment result.
+     *      - Proactively unschedule the worker.
+     * 2. It fails to emit a {@link WorkerLaunched} event.
+     *      - The worker will get unscheduled for this launch task request.
+     * 3. There are no launch tasks for this assignment result.
+     *      - All of these leases are rejected.
+     *      - Eventually, the underlying Mesos driver will decline offers since there are no launch task requests.
+     *
+     * @param requests collection of assignment results received by the scheduler.
+     * @param leases list of resource offers from Mesos.
+     */
     private void launchTasks(Collection<TaskAssignmentResult> requests, List<VirtualMachineLease> leases) {
         List<LaunchTaskRequest> launchTaskRequests = new ArrayList<>();
+
         for (TaskAssignmentResult assignmentResult : requests) {
             ScheduleRequest request = (ScheduleRequest) assignmentResult.getRequest();
-            WorkerPorts workerPorts = new WorkerPorts(assignmentResult.getAssignedPorts());
-            boolean success = jobMessageRouter.routeWorkerEvent(new WorkerLaunched(
-                    request.getWorkerId(),
-                    request.getStageNum(),
-                    leases.get(0).hostname(),
-                    leases.get(0).getVMID(),
-                    getAttribute(leases.get(0), slaveClusterAttributeName),
-                    workerPorts));
-            if (success) {
-                launchTaskRequests.add(new LaunchTaskRequest(request, workerPorts));
+
+            WorkerPorts workerPorts = null;
+            try {
+                workerPorts = new WorkerPorts(assignmentResult.getAssignedPorts());
+            } catch (IllegalArgumentException | IllegalStateException e) {
+                logger.error("problem launching tasks for assignment result {}: {}", assignmentResult, e);
```

Comment:
Since this block is hit primarily for invalid /missing ports, should we name the counter appropriately so we can track all invalid port related issues together?

Reply:
Yeah that works for me. I'll use that same counter `numMissingWorkerPorts`.